Cython/Compiler/Nodes.py

```diff
@@ -2267,6 +2273,17 @@ def assure_gil(code_path, code=code):
                     code.putln("%s = %s;" % (Naming.retval_cname, err_val))
             elif not return_type.is_void:
                 code.putln("__Pyx_pretend_to_initialize(&%s);" % Naming.retval_cname)
+                # For C, __Pyx_pretend_to_initialize() is sufficient (taking an address
+                # is enough to make returning a variable defined). For C++ this isn't
+                # true, and we genuinely have to make sure a variable is initialized.
+                empty_decl_code = return_type.empty_declaration_code()
+                # C++ doesn't want names of the form '= struct foo()' so
+                # remove the struct
+                empty_decl_code = empty_decl_code.replace("struct ", "").replace("union ", "")
```

Comment:
I'm trying to think of situations where this simple replacement text might not be safe. What about function pointers, for example? Couldn't those have spaces after user defined type names, like a typedef name `sometypestruct` ending in `…struct ` with a space?

Reply:
I think a function pointer would end to using `NULL` to signal an error so wouldn't get used by the "uninitialised return on error" path.

However, I take the general point here - I'll have another look. Worst case would be to restrict the replacement to the start of strings...

I think I've also missed `enum ` although they might also have a well-defined error return value.

(I think this change can miss the next beta though)

Reply:
I've done this in PyrexTypes instead, which is a bit more verbose but should be more reliable